Current workout page

HugoGoHe opened the pull request:
The current workout page was implemented as well as the Exercise object that is being stored in the Parse database.

I also worked on the feature to show the suggestions when the user is typing the exercises. This is being done with a textfield and a second tableview.

https://user-images.githubusercontent.com/88735772/180590321-7c926efa-6b9c-405e-8560-a0a9ba7a37c9.mov

HugoGoHe committed:
Merge remote-tracking branch 'origin/Workouts_page' into Workouts_page

HugoGoHe committed:
Auto stash before merge of "Workouts_page" and "origin/Workouts_page"

HugoGoHe committed:
[New PFObject subclass "Exercise"]

Created .m and .h files for the class Exercise.

HugoGoHe committed:
[ExerciseCell]

Created .m and .h files for the class of the prototype cell of the tableView to show the exercises.

HugoGoHe committed:
[Current Workout page][getExercises & refreshControl]

Implemented the method getExercises to query the data and pull to refresh feature.

HugoGoHe committed:
[Current Workout page] [cellForRowAtIndexPath]

cellForRowAtIndexPath method

HugoGoHe committed:
[Current Workout page][Adding exercise]

Now the user can add exercises in the workouts.

HugoGoHe committed:
[Current Workout page][Validating user input]

Now the user input is being validated showing an error message so that erroneous data is not saved . Also when data is submitted the text inside the textfields is deleted.

- Validate that the text fields are not empty
- Validates that the weight entered is a number and that it is greater than 0
- Validates that the sets input has the correct format

HugoGoHe committed:
[Current Workout page][Deleting rows from table view and Parse]

User can delete exercises from the tableView and are also deleted from the Parse database.

HugoGoHe committed:
[Current Workout page][Deleting rows from table view and Parse]

User can delete exercises from the tableView and are also deleted from the Parse database.


BUG:
The program sometimes crashes trying to load the exercises of a previos workout without creating a new workout before.

HugoGoHe committed:
Merge remote-tracking branch 'origin/CurrentWorkoutG' into CurrentWorkoutG

HugoGoHe committed:
[Current Workout List][List of available exercises]

Replaced the searchbar with a textfield and created a second table view that shows the available exercises and is updated with the user's typing, showing the exercises that contain the substring of the textfield.

HugoGoHe committed:
[Current Workout Page][Selecting an exercise]

- The tableView is hidden when te user finishes editing.
- The tableView is displayed with every exercise when the user begins editing.
- When the user presses a cell, the exercise contained in it is autocompleted in the textfield.
- User input is validated, if the exercise is not contained in the array of exercises an error label appears.

HugoGoHe committed:
[Current Workout][UI Changes][Fixed BUG}

- Changes in the UI
- Changed the name of properties that started with the word "set"
- Now the list of available exercises doesn't shows up when user begins editing the exerciseTextField
- Changed @dynamic to @synthesize on Exercise.m:
       -This fixed the bug that crashed the app when trying 
       to load the exercises table view of past exercise.
- Observer for the size of the autocompleteTableView: now the table of suggestions of available exercises is being resized when the text on the textfield changes.
- Workouts tableView row height was changed.


- Cleaned the code

HugoGoHe committed:
[Current Workout page][@synthesize to @dynamic]

Had to change again @synthesize to @dynamic in the properties of the Exercise class because the exercises were not being uploaded to the Parse database.

Now the bug that was occurring because of this doesn't  happen anymore.

HugoGoHe committed:
[Stats][Adding HighCharts framework]

- Added the pod
- Included the framework

HugoGoHe committed:
[Stats page][ Loading test charts]

Wrote the code to display simple line chart in view and tableView.

HugoGoHe committed:
[Current workout page][Suggested exercises optimization]

Taking advantage of cached content by making a query every time there is only one letter in the string.

For example, a user is typing 'B'. Then all the exercises starting with B is fetched. Then the user types e ('Be'). Here since 'Be' results is a subset of ' B' results. It filters the data from the local cache and this will continue for subsequent chars until user deletes to whole string and starts from scratch.

cmnagarajan commented on Capstone Project/PFObjects/Exercise.m:
```diff
@@ -10,13 +10,13 @@
-@synthesize exerciseSet4;
-@synthesize exerciseSet5;
-@synthesize workout;
+@dynamic weight;
```
If this is dynamic then the setters and getters are not auto generated
You have to add them explicitly. Please dont proceed just because it is working somehow

-(void)setExerciseSet1:(NSInteger)value
{
  self.excerciseSet1 = value;
}

- (NSInteger)exerciseSet1
{
  return exerciseSet1;
}

HugoGoHe replied:
"No need if it works without that as documented"

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -33,12 +31,17 @@ - (void)viewDidLoad {
     self.exerciseTextField.delegate = self;
     
-    self.arrayOfExercises = [[NSMutableArray alloc] init];
+    self.autocompleteTableView.delegate = self;
```
This is suggested text. Autocomplete shows up on the same text field (as it is in some chat apps). Suggested text offers multiple options as drop down whereas automcomplete offers only one option. Please rename this to suggestedTextTableView

cmnagarajan commented on Capstone Project/PFObjects/Exercise.m:
```diff
@@ -27,11 +27,11 @@ +(void) newExercise: (NSString *_Nullable)name withWeight:(float)weight withSet1
-    newExercise.set3 = set3;
-    newExercise.set4 = set4;
-    newExercise.set5 = set5;
+    newExercise.exerciseSet1 = set1;
```
This can be an array or a formatted string (x/x/x/x) stored the same way in db too. Only when it needs to be rendered in UI as labels, it can use '/' delimiter API to load in an array and set on labels.

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -232,6 +246,13 @@ - (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEd
+    }
+}
+
 
```
The add button is always enabled. It should be enabled only when there are all required entries with valid data

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -232,6 +246,13 @@ - (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEd
+        self.exerciseTextField.text = [self.autocompleteExercises objectAtIndex:indexPath.row];
+    }
+}
+
```
Center the exercise field and suggestedTextTableView in the UI .
Remove the blue line in the table view

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -84,14 +84,65 @@ -(void)getExercises{
-            [self getExercises];
+    //Check for any blank fields
+    if([self.searchBar.text isEqual:@""] || [self.weightTextField.text isEqual:@""] || [self.repsTextField.text isEqual:@""]){
+        self.errorLabel.text = @"One or more fields are Blank";
```
self.errorLabel.text = @"One or more fields are blank";

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -84,14 +84,65 @@ -(void)getExercises{
+        NSArray *repsPerSet = [self.repsTextField.text componentsSeparatedByString:@"/"];
+        //Checking if sets are valid
+        if(![self areSetsValid:repsPerSet]){
+            self.errorLabel.text = @"Input not valid";
```
Sets input is not valid

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -84,14 +84,65 @@ -(void)getExercises{
+            while(repsPerSetMut.count < 5){
+                [repsPerSetMut addObject: @"0"];
+            }
+            [Exercise newExercise:self.searchBar.text withWeight:[self.weightTextField.text floatValue] withSet1:[[repsPerSetMut objectAtIndex:0] intValue] withSet2:[[repsPerSetMut objectAtIndex:1] intValue] withSet3:[[repsPerSetMut objectAtIndex:2] intValue] withSet4:[[repsPerSetMut objectAtIndex:3] intValue] withSet5:[[repsPerSetMut objectAtIndex:4] intValue] withWorkout:self.selectedWorkout wi ...
```
What if the size of the array is less than 5. The number of sets done on that day is less. If that happens objectAtIndex:4 call will results in crash. Using it without knowing the actual size of the array is a common cause for crash

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -84,14 +84,65 @@ -(void)getExercises{
+}
+
+-(BOOL) areSetsValid:(NSArray*) repsPerSet{
+    int maxNumSets = 5;
```
Declare this as static const value at the top of the file instead of a local variable

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -84,14 +84,65 @@ -(void)getExercises{
 }
 
+-(BOOL) isSetValid:(id) set{
+    int maxNumReps = 99;
```
same as above

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -57,11 +57,11 @@ - (IBAction)didTapSave:(id)sender {
     [exerciseQuery whereKey:@"workout" equalTo:self.selectedWorkout];
     [exerciseQuery orderByDescending:@"createdAt"];
     [exerciseQuery findObjectsInBackgroundWithBlock:^(NSArray<Exercise *> * _Nullable exercises, NSError * _Nullable error) {
+        NSLog(@"%@", exercises);
```
Use lldb to debug to avoid using NSLog in the code

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -148,18 +147,27 @@ - (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForR
+    cell.set2Label.text = [NSString stringWithFormat:@"%d", exercise.set2];
+    cell.set3Label.text = [NSString stringWithFormat:@"%d", exercise.set3];
+    cell.set4Label.text = [NSString stringWithFormat:@"%d", exercise.set4];
+    cell.set5Label.text = [NSString stringWithFormat:@"%d", exercise.set5];
```
what happens when the number of sets is less than 5

HugoGoHe replied:
![image](https://user-images.githubusercontent.com/88735772/181415382-80358a2b-4316-44e8-9d7e-9054575bf105.png)

cmnagarajan commented on Capstone Project/ViewControllers/Workouts/CurrentWorkoutViewController.m:
```diff
@@ -89,10 +80,17 @@ -(void)getExercises{
 
+- (IBAction)didTapAdd:(id)sender {
+    NSArray *repsPerSet = [self.repsTextField.text componentsSeparatedByString:@"/"];
+    [Exercise newExercise:self.searchBar.text withWeight:[self.weightTextField.text floatValue] withSet1:[[repsPerSet objectAtIndex:0] integerValue] withSet2:[[repsPerSet objectAtIndex:1] integerValue] withSet3:[[repsPerSet objectAtIndex:2] integerValue] withSet4:[[repsPerSet objectAtIndex:3] integerValue] withSet5:[[repsPerSet objectAtIndex:4] integerValue] withWorkout:self.selectedWorkout withC ...
```
Directly accessing index without checking the size of the array is a recipe for crash

HugoGoHe committed:
[Current workout page][Refactor]

Naga's review on pr:

- Now sets are not individual properties, instead they are stored in an NSArray property called exerciseSets.

- Renamed variables with prefix "autocomplete" to "suggested".

- Corrected typo in error message for empty fields.

- Changed error message for sets input.

- Declared static constants por maximum number of reps and maximum number of sets.

- Centered the exerciseTextField and removed the blue separation lines from suggestedTextTableView.

- SOLVED BUG:
Now the query for the available exercises is done when the user types the first letter in the exerciseTextField. Before it was being done with an empty character ""

HugoGoHe committed:
[Stats page][CurrentWorkout page][ Refactored code from pr's feedback]

Removed arrayOfExercises allocation and commented code from the test charts.

HugoGoHe committed:
[Current workout page][Index error]

The text of every label was being replaced with the object at index 0, instead of its corresponding index

HugoGoHe committed:
Merge branch 'Stats_page' into CurrentWorkoutG
